Fix invalid race condition check

milljm opened the pull request:
Iterate over all runnable jobs and discover potential
race conditions.

Closes #11454

Commit:
Fix invalid race condition check

Iterate over all runnable jobs and discover potential
race conditions.

Closes #11454

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/188752/) on 570e043 wanted to post the following:

View the site [here](http://mooseframework.inl.gov/docs/PRs/11483/site)

This comment will be updated on new commits.

permcody commented on python/TestHarness/JobDAG.py:
```diff
@@ -128,26 +128,26 @@ def _doSkippedDependencies(self):
+            for job in list(job_list):
+                for match_job in self.__job_dag.all_downstreams(job):
+                    if match_job in job_list:
+                        job_list.remove(match_job)
```
Good! This looks like the correct algorithm.

milljm replied:
But it also is nested deep in for loops :/ I was honestly hoping you had some magic python up your sleeve to beautify this!

permcody replied:
If you use sets, you can remove the inner loop directly. Look, you are just subtracting matching jobs there, so it's just `job_list = job_list - match_jobs`.

milljm replied:
@permcody I could do this... but then I would have to store the jobs contained in output_to_job dictionary as a set, instead of a list (or add more lines to convert it later).

What we talked about before, using sets might be slower, and will allow syntax mistakes like this:
https://github.com/idaholab/moose/blob/380145aacd2d1d2f32485bb2cc16b497fc0633ad/test/tests/mesh/adapt/tests#L5

I like the idea of sets. But I thought we agreed upon lists...?

permcody replied:
Stick with lists, I bet the sets are slower (for small sets)

permcody replied:
OK to clarify.... Sets are always slower than lists, but your algorithm as it stands is `O(m * n * o)` but by using lists you could make it `O(m * (n + o))`. However, the overhead for these small sets would make the second complexity larger than the first.

milljm replied:
Thats odd, I went ahead and used sets just for curiosity's sake, removed the inner loop. But then received a race condition with ./run_tests -i duplicate_outputs_ok with tests i and j.

I'll use lists as we discussed.

The error was strange:
```pre
    FAIL tests/test_harness.h FAILED (OUTFILE RACE CONDITION)
    FAIL tests/test_harness.i FAILED (OUTFILE RACE CONDITION) [skipped dependency]
```

Race condition with a skipped dependency caveat... Ummm...

permcody replied:
Did you use sets for both matched_job and job? If not the subtraction operator might still work but it might fail. You need two sets.

permcody reviewed (changes requested):
This looks good except you didn't add a new test case (e.g. the one one the board). That test case doesn't work with the old logic but should with this logic. Let's get it in there so we can verify this works.

Commit:
Add unittest to test delayed race conditions

Closes #11454

Commit:
Handle multiple duplicate files

Handle tests which claim to touch the same output file multiple times.
We will treat this new behavior as an error.

Closes #11454
